### PR TITLE
WIP: dojson: support enhanced marcxml links

### LINKED ISSUE #485: dojson: support for enhanced MARCXML links
https://github.com/inspirehep/inspire/commit/de148e839653c18ad0c6572874ffe85702509482 and followed up commits have introduced several links across records.
```    
         100/700 $x with Record ID of linked HepName,
                 $y with True/False if the signature is claimed
                 $z with Record ID of institution
         371/110 $z with Record ID of institution
         502     $z with Record ID of institution
         999C5   $0 with on the fly discovered Record IDs (not for books)
         773     $0 with Record ID of corresponding Book or Proceeding
                 $1 with Record ID of corresponding Journal
                 $2 with Record ID of corresponding Conference
         693/710 $0 with Record ID of corresponding experiment
```

Dojson scripts should amended accordingly to translate this links from MARCXML into JSON.
* [x] `100/700 $x` in HEP should go to `authors[].recid`
* [x] `100/700 $y` in HEP should go to `authors[].curated_relation` (in `dojson` this is currently called `claimed`)
* [x] `100/700 $z` in HEP should go to `authors[].affiliations[].recid` (`curated_relation` is set to `true` if there is a recid.
* [x] `502 $z` in HEP should go to `thesis[].recid` (`curated` blablabla)
* [x] `371 $z` in HepNames should go to `positions[].institution.recid` (`curated_relation to `true`)
* [x] `110 $z` in Jobs should go to `institution.recid` (`curated_relation to `true`)
* [x] `999C5  $0` in HEP should go into `references[].recid`
* [x] `773 $0` in HEP should go into into new `book.recid` field if `BOOKCHAPTER`
* [x] `773 $1` in HEP should go into `publication_info.journal_recid` (which should be added to `hep-0.0.1.json`
* [x] `773 $2` in HEP should go into `publication_info.recid`
* [x] `693 $0`in HEP should go to `accelerator_experiment.recid` (`curated_relation` set to `true`)
* [x] `710 $0`in HEP should go to `collaboration.recid` (does not exist yet.)

### PR DESCRIPTION
**Pull request by ioannistsanaktsidis**
* Closes #485.

Signed-off-by: Ioannis Tsanaktsidis <ioannis.tsanaktsidis@cern.ch>

### CODE REVIEW

**Comment by kaplun** [comment]
One regression test to go :+1:

**Comment by kaplun** [comment]
Please, be more verbose in the commit log, and explain the changes (imagine github is down :smile: ) You can cut and paste all the details on MARC :)

**Comment by ioannistsanaktsidis** [comment]
I addressed all the bullets from the issue. Concerning this one 773 $0 in HEP should go into into new book.recid field if BOOKCHAPTER I have implemented something in model.py. It seems to work for me but I am not about 100% abou the way I do it. :/

**Comment by jmartinm** [comment]
@ioannistsanaktsidis There are still a couple of tests failing.

**Comment by ioannistsanaktsidis** [comment]
Yes I know, I will check it today so as to complete it.

File: inspirehep/base/templates/format/record/Inspire_Default_HTML_brief.tpl [hunk 1]
```diff
@@ -51,11 +51,9 @@
                     {% endif %}
                   </b>
                 </h4>
-                {% if record.authors %}
```
Comment (jmartinm) [inline inspirehep/base/templates/format/record/Inspire_Default_HTML_brief.tpl]: Seems a bit unrelated to this PR, this if is not needed?

Comment (ioannistsanaktsidis) [inline inspirehep/base/templates/format/record/Inspire_Default_HTML_brief.tpl]: I changed this and the general_macros file due to the changes I made in the collaboration field so as not to break it. Do you think we should move it to another PR?

File: inspirehep/dojson/hep/fields/bd70x75x.py [hunk 2]
```diff
@@ -57,7 +57,13 @@ def thesis_supervisor2marc(self, key, value):
 def collaboration(self, key, value):
     """Added Entry-Corporate Name."""
     def get_value(value):
-        return value.get('g')
+        recid = ''
+        if '0' in value:
+            recid = int(value.get('0'))
```
Comment (jmartinm) [inline inspirehep/dojson/hep/fields/bd70x75x.py]: This will crash if there is no key `0` in value. You can do the `try...except` like above.

File: inspirehep/dojson/hep/fields/bd76x78x.py [hunk 3]
```diff
@@ -41,17 +42,20 @@ def publication_info(self, key, value):
             # Some crap in they year :-(
             pass
     try:
-        if '0' in value:
-            if isinstance(value['0'], list):
-                recid = int(value['0'][0])
+        if '2' in value:
+            if isinstance(value['2'], list):
```
Comment (jmartinm) [inline inspirehep/dojson/hep/fields/bd76x78x.py]: It's better to access all values with `value.get` to avoid crashing.

File: inspirehep/dojson/hep/model.py [hunk 4]
```diff
@@ -26,4 +26,19 @@
 from dojson import utils
 
 hep = Overdo()
+
+
+@hep.over('book', '^773..')
```
Comment (jmartinm) [inline inspirehep/dojson/hep/model.py]: Looking at the `dojson` source code I don't think this ensures that this code will always be run at the end (after the collections have been populated)

With a bit of Python magic, I have created this PR (https://github.com/inspirehep/inspire-next/pull/666) to show another way that would ensure we execute the code at the end.

@kaplun @jalavik what do you think? This is in order to allow for some code to execute after dojson is finished (to ensure all the keys are present)

Comment (ioannistsanaktsidis) [inline inspirehep/dojson/hep/model.py]: @jmartinm  you are right, thats what I wanted to discuss also. This thing that I did is not running for every record. Sometimes it is missing.

**Comment by ioannistsanaktsidis** [comment]
All tests are passing. We can wait for @jmartinm  PR to be merged and then we can merge this one.

File: inspirehep/dojson/hep/model.py (hunk 4 above)
Comment (jalavik) [inline inspirehep/dojson/hep/model.py]: Yeah, the order is not guaranteed. In the past we would add the extra stuff directly in the function (in this case collections), but it is not always nice as collections is now shared between all dojsons in `base.py`.

@jmartinm Another possibly less "magic" way could be to use inheritence to subclass the `Overdo` class in the hep model and there override the `do()` function to call the super version and then do our code.

Comment (jmartinm) [inline inspirehep/dojson/hep/model.py]: Good point @jalavik Actually did not read this comment in time and merged https://github.com/inspirehep/inspire-next/pull/666 after a slight modification that allows to keep using `hep.do`

The effect is the same, so at least somebody can learn how to inject a function into a class instance :)

Comment (ioannistsanaktsidis) [inline inspirehep/dojson/hep/model.py]: I guess this someone is me :P :D

**Comment by jmartinm** [comment]
Continues here https://github.com/inspirehep/inspire-next/pull/674
